cmd/containerboot/main.go

```diff
@@ -0,0 +1,426 @@
+		// zombies.
+		for {
+			var status unix.WaitStatus
+			pid, err := unix.Wait4(-1, &status, 0, nil)
+			if err != nil {
+				log.Fatalf("Waiting for exited processes: %v", err)
+			}
+			if pid == daemonPid {
+				log.Printf("Tailscaled exited")
+				os.Exit(0)
+			}
+		}
+	}()
+
+	// Wait for the socket file to appear, otherwise 'tailscale up'
+	// can fail.
+	log.Printf("Waiting for tailscaled socket")
+	for {
+		if ctx.Err() != nil {
+			log.Fatalf("Timed out waiting for tailscaled socket")
+		}
+		_, err := os.Stat(cfg.Socket)
+		if errors.Is(err, fs.ErrNotExist) {
+			time.Sleep(100 * time.Millisecond)
+			continue
+		} else if err != nil {
+			log.Fatalf("Waiting for tailscaled socket: %v", err)
+		}
+		break
+	}
+
+	if !cfg.AuthOnce {
+		if err := tailscaleUp(ctx, cfg); err != nil {
+			return nil, fmt.Errorf("couldn't log in: %v", err)
+		}
+	}
+
+	tsClient := tailscale.LocalClient{
+		Socket:        cfg.Socket,
+		UseSocketOnly: true,
+	}
+
+	// Poll for daemon state until it goes to either Running or
+	// NeedsLogin. The latter only happens if cfg.AuthOnce is true,
+	// because in that case we only try to auth when it's necessary to
+	// reach the running state.
+	for {
+		if ctx.Err() != nil {
+			return nil, ctx.Err()
+		}
+
+		loopCtx, cancel := context.WithTimeout(ctx, time.Second)
+		st, err := tsClient.Status(loopCtx)
+		cancel()
+		if err != nil {
+			return nil, fmt.Errorf("Getting tailscaled state: %w", err)
+		}
+
+		switch st.BackendState {
+		case "Running":
+			if len(st.TailscaleIPs) > 0 {
+				return st, nil
+			}
+			log.Printf("No Tailscale IPs assigned yet")
+		case "NeedsLogin":
+			// Alas, we cannot currently trigger an authkey login from
+			// LocalAPI, so we still have to shell out to the
+			// tailscale CLI for this bit.
+			if err := tailscaleUp(ctx, cfg); err != nil {
+				return nil, fmt.Errorf("couldn't log in: %v", err)
+			}
+		default:
+			log.Printf("tailscaled in state %q, waiting", st.BackendState)
+		}
+
+		time.Sleep(500 * time.Millisecond)
+	}
+}
+
+// tailscaledArgs uses cfg to construct the argv for tailscaled.
+func tailscaledArgs(cfg *settings) []string {
+	args := []string{"--socket=" + cfg.Socket}
+	switch {
+	case cfg.InKubernetes && cfg.KubeSecret != "":
+		args = append(args, "--state=kube:"+cfg.KubeSecret, "--statedir=/tmp")
+	case cfg.StateDir != "":
+		args = append(args, "--state="+cfg.StateDir)
+	default:
+		args = append(args, "--state=mem:", "--statedir=/tmp")
+	}
+
+	if cfg.UserspaceMode {
+		args = append(args, "--tun=userspace-networking")
+	} else if err := ensureTunFile(); err != nil {
+		log.Fatalf("ensuring that /dev/net/tun exists: %v", err)
+	}
+
+	if cfg.SOCKSProxyAddr != "" {
+		args = append(args, "--socks5-server="+cfg.SOCKSProxyAddr)
+	}
+	if cfg.HTTPProxyAddr != "" {
+		args = append(args, "--outbound-http-proxy-listen="+cfg.HTTPProxyAddr)
+	}
+	if cfg.DaemonExtraArgs != "" {
+		args = append(args, strings.Fields(cfg.DaemonExtraArgs)...)
```

Comment:
`strings.Fields` is ... technically not the same as before, since it doesn't do shell escaping in the same way (e.g. `TS_TAILSCALED_EXTRA_ARGS="foo \"bar baz\"" ./run.sh` behaves differently now). I'm happy to ignore if you are, though.

Reply:
Oh this is annoying :(

Reply:
Looking at the args for both tailscaled and tailscale up, I don't see any reasonable commandline you could construct that embeds quoted elements with spaces. On that basis I don't feel strongly about adding a bunch of complexity here. WDYT?

Reply:
I was thinking someone wanted to pass a `--hostname` or `--operator` with a value that has spaces; something like this: `TS_TAILSCALED_EXTRA_ARGS='--hostname "with spaces"'`, which would previously (correctly) set the hostname to `"with spaces"` and now... presumably just passes garbage?

I'm happy to say "don't do that" here, though.

Reply:
SGTM